lib/Sema/CSGen.cpp

```diff
@@ -1227,9 +1227,11 @@ namespace {
         if (knownType) {
```

Comment:
Just above this, within the `CS.getVarType(VD)` call it's getting the type of VD and replacing any ErrorTypes it contains with holes when CS.isForCodeCompletion is true. Does this change make that logic redundant?

Reply:
I think for code completion we'd want to preserve as much structure as possible so replacing everything with a single hole might not be the best way... I was thinking more of a second case where `E->getDecl()->isInvalid()` is true.

Reply:
If it is, could we remove it and make the below behave the same way, preserving any wrapping types? (e.g. `[ErrorType]` becomes `[<hole>]` rather than just `<hole>`?

Reply:
Didn't see your reply before I sent that, sorry.

Reply:
I guess I could try to unify logic here to preserve some more structure for diagnostics, I'll give it a try!

Reply:
I looked into this some more and it looks like for solver and diagnostics in particular it make more sense to keep invalid declarations represented as type variables so they could assume any time from context instead otherwise we'd risk diagnosing issues which are effectively already diagnosed while validating declarations...